### DISCUSSION TITLE
Server-Side Uncaught Error Handler

### DISCUSSION THREAD

**Discussion by sergiodxa** [remix-run/remix#2190] [Proposals]
# Proposal

Add an optional export in `entry.server` to receive any uncaught error (errors that triggered an ErrorBoundary) in a centralized place.

## Background

Right now, Remix catches any error throw by the app to render the matching ErrorBoundary.

That is great but what if I want to send server-side errors to Sentry or something similar?

Right now I had to wrap the loader/action code in a try/catch, catch the error, send it to Sentry and throw the error again (supposing I don't want to handle it and throw a response).

And that only works on loader/action, I will not be able to catch render errors (maybe on entry.server?) because Sentry needs a different SDK for client and server, and since the ErrorBoundary code runs in both sides I would had to initialize both and detect where I'm to use one or the other.

## Use Case

- Centralize server-side error reporting.
- Provide a place for 3P services like Sentry, Bugsnag, etc. to add uncaught error reporting.

## API/Example

In `entry.server` add an optional export

```ts
export let handleUncaughtError: HandleUncaughtErrorFunction = async ({ request, error, context }) => {
  Sentry.captureException(error);
}
```

The function could also receive the request object that had the error so we could get info from there to send to Sentry.

With a function like this, I could have a way to know any error I didn't specifically caught, I will still not be able to intercept the response so error boundaries will continue to work but still get proper error reporting.

The idea is the return value of the function is completely ignored, it's an async function only so you can do async work there like parsing the session from the request.

## Prior Art

- @kiliman has implemented something like this here https://github.com/kiliman/remix-error-logging

**Comment by kiliman** [comment]
Yes, I was going to suggest something like this. Currently Remix simply logs to `console.error` on server and returns a generic `500 Unexpected Server Error` on production.

It would definitely be helpful to include a hook to be able to log server errors.

**Comment by dcramer** [comment]
Dropping this here in case anyone else stumbles upon this thread: https://twitter.com/mjackson/status/1515351242637152262

**Comment by sergiodxa** [comment]
This patch from @kiliman looks to do exactly what I describe here https://github.com/kiliman/remix-error-logging

**Reply by kiliman** [reply]
Interesting how @jacob-ebey approached the Sentry integration. He wrapped each route module with the Sentry function. Granted he wasn't patching Remix, but one thing I haven't played with is how to include context in the bug report, like current user. I want to make sure the proposed API will work for different use cases before submitting a PR.

**Reply by sergiodxa** [reply]
Since you have access to the error on the `handleError` function, you could fetch/query the user from the auth cookie (if defined) and add it to the Sentry context.

**Reply by kiliman** [reply]
I've updated my patch to support per request context using `getLoadContext`.

**Reply by joshboley** [reply]
Does anyone know how to get this working with newer versions of Remix?  Looks like the patches only support 1.6.x.

**Reply by kiliman** [reply]
I'll try to get my patch updated for latest.

**Reply by kiliman** [reply]
Patch updated for Remix v1.16.1

https://github.com/kiliman/remix-error-logging/blob/main/patches/%40remix-run%2Bserver-runtime%2B1.16.1.patch

**Reply by quinn** [reply]
@sergiodxa Any chance we could get this patch into v2?

**Reply by kiliman** [reply]
There's a PR to add `handleError`. Not sure if it'll make it to v2, but @brophdawg11 wrote it so chances are likely.

https://github.com/remix-run/remix/pull/6495

**Reply by kiliman** [reply]
I updated my patch to match the PR, so now the `handleError` has the same signature:

```ts
export interface HandleErrorFunction {
  (error: unknown, args: DataFunctionArgs): void;
}
```

**Comment by peter-unify** [comment]
I think Remix should provide more hooks for those that don't want to "eject" to express just to handle things like middleware and proper error handling. The DX of remix-serve is much better than than the express + remix setup.

I think adding middleware support for loaders/actions would go a long way here - it would be trivial to hook into the middleware to add performance logging, error reporting, etc. if we had that capability. Right now it seems the recommendation is to just use express, which is unfortunate.

**Comment by quinn** [comment]
I would love to have this. Not sure why, but a standard Express error handler middleware is not firing. I would like to log errors so I can be alerted on the error. Also, the current Remix setup just dumps the error into STDERR, which isn't supported by my log pipeline (all logs must be JSON)

**Comment by sergiodxa** [comment]
This is now implemented in [v1.17.0](https://github.com/remix-run/remix/releases/tag/remix%401.17.0).
